File: src/Random/Random.js
/*******************************************************************************
 *  Copyright by the contributors to the Dafny Project
 *  SPDX-License-Identifier: MIT
 *******************************************************************************/

var DafnyLibraries = DafnyLibraries || {};
DafnyLibraries.Random = (function () {
    let $module = {};

    $module.nextBool = function () {
        try {
        return Math.random() < 0.5;
        } catch (e) {
            throw "nextBool failed";
        }
    }

    $module.nextInt = function (bound) {
        bound = bound > 0 ? bound : Number.MAX_SAFE_INTEGER;
        return new BigNumber(Math.floor(Math.random() * bound));
    }

    $module.nextReal = function () {
        num = Math.floor(Math.random() * Number.MAX_SAFE_INTEGER);
        den = Number.MAX_SAFE_INTEGER;
        return new _dafny.BigRational(new BigNumber(num), new BigNumber(den));
    }

    return $module;
})();

File: src/Random/Random.java
/*******************************************************************************
 *  Copyright by the contributors to the Dafny Project
 *  SPDX-License-Identifier: MIT
 *******************************************************************************/

package DafnyLibraries;

import dafny.*;
import java.math.*;

public class Random {
    public static boolean nextBool()
    {
        return Math.random() < 0.5;
    }
    public static BigInteger nextInt(BigInteger Bound)
    {
        long bound = Bound.longValue();
        long bnd = bound > 0 ? bound : Long.MAX_VALUE;
        return BigInteger.valueOf((long) Math.floor(Math.random() * bnd));
    }
    public static BigRational nextReal()
    {
        long num = (long) Math.floor(Math.random() * Long.MAX_VALUE);
        long den = Long.MAX_VALUE;
        return new BigRational(BigInteger.valueOf(num), BigInteger.valueOf(den));
    }
}
